482p3/man/man8/hgfs.8
.TH HGFS 8
.SH NAME
hgfs \- VMware Host/Guest File System server
.SH SYNOPSIS
\fBmount \-t hgfs \fR[\fB\-r\fR] [\fB\-o \fIoptions\fR] \fBnone \fImountpoint
.SH DESCRIPTION
The Host/Guest File System (HGFS) server allows one to mount
VMware Shared Folders as a file system. This makes it possible to access
selected portions of the VMware host file system when MINIX is run as a
VMware guest.
.PP
The above mount command will mount the hgfs file system onto the directory
\fImountpoint\fR. The \fB\-r\fR mount option makes the file system read-only;
note that shared folders may independently have been configured as read-only
on the VMware host. The \fIoptions\fR field is a string consisting of
comma-delimited \fIkey\fR or \fIkey\fB=\fIvalue\fR options. The following
options are supported.
.TP 4
\fBprefix=\fIpath\fR
This option sets a path prefix that will be prepended to all file system
operations on the host system. When mounted without a prefix (the default),
the root directory of an HGFS mount will contain all the names of the
available shares. The prefix option can be used to mount one of those shares,
by specifying its name as the prefix. Multi-component path prefixes are
supported as well.
.TP
\fBuid=\fInumber\fR
This sets the user ID used for all the files and directories in the file
system, allowing a non-root user to be the owner. The value must be specified
as a decimal number.
The default is root (the number \fB0\fR).
.TP
\fBgid=\fInumber\fR
Likewise, sets the group ID for all files and directories.
The default is operator (the number \fB0\fR).
.TP
\fBfmask=\fInumber\fR
This option sets the file permission mask of regular files. It is specified as
an octal number. For example, a value of \fB600\fR makes all files readable and
writable by the owning user (see the "\fBuid\fR" option).
The default is \fB755\fR.
.TP
\fBdmask=\fInumber\fR
Likewise, sets the file permission mask of directories.
The default is also \fB755\fR.
.TP
\fBicase\fR
This option tells HGFS to treat names as case-insensitive.
.TP
\fBnoicase\fR
This option, set by default, reverts the effect of an earlier specified
"\fBicase\fR" option.
.SH EXAMPLES
.TP 20
.B mount \-t hgfs none /mnt
# Mount the entire shared folders tree on \fI/mnt\fR
.TP 20
.B mount \-t hgfs \-o prefix=shared,uid=20,fmask=644,icase none /usr/shared
# Mount the "\fIshared\fR" shared folder on \fI/usr/shared\fR
.SH LIMITATIONS
HGFS uses the first and original version of the VMware Shared Folders protocol
to talk to the VMware host. That means that HGFS should work with all VMware
products that support shared folders. However, this also imposes a number of
limitations. For example, the first version of the protocol supports only
regular files and directories (and not links), and does not have support for
automatic case sensitivity handling.
.PP
Some file system operations may not behave as expected, because the behavior
of HGFS is determined largely by the host. Other file system operations
(in particular, using directories as mountpoints) are not implemented,
because the file system structure as perceived by HGFS may change arbitrarily
at any time, due to modifications on the host side.
.SH "SEE ALSO"
.BR mount (1),
.BR vbfs (8)
.SH AUTHOR
David van Moolenbroek <dcvmoole@cs.vu.nl>


482p3/servers/hgfs/hgfs.c
/* This file contains the implementation of the HGFS file system server.
 * The file system side is handled by libsffs, whereas the host communication
 * is handled by libhgfs. This file mainly contains the glue between them.
 *
 * The entry points into this file are:
 *   main		main program function
 *
 * Created:
 *   April 2009 (D.C. van Moolenbroek)
 */

#include <minix/drivers.h>
#include <minix/sffs.h>
#include <minix/hgfs.h>
#include <minix/optset.h>

static struct sffs_params params;

static struct optset optset_table[] = {
  { "prefix",   OPT_STRING, params.p_prefix,       sizeof(params.p_prefix) },
  { "uid",      OPT_INT,    &params.p_uid,         10                      },
  { "gid",      OPT_INT,    &params.p_gid,         10                      },
  { "fmask",    OPT_INT,    &params.p_file_mask,   8                       },
  { "dmask",    OPT_INT,    &params.p_dir_mask,    8                       },
  { "icase",    OPT_BOOL,   &params.p_case_insens, TRUE                    },
  { "noicase",  OPT_BOOL,   &params.p_case_insens, FALSE                   },
  { NULL,       0,          NULL,                  0                       }
};

/*===========================================================================*
 *				sef_cb_init_fresh			     *
 *===========================================================================*/
static int sef_cb_init_fresh(int UNUSED(type), sef_init_info_t *UNUSED(info))
{
/* Initialize this file server. Called at startup time.
 */
  const struct sffs_table *table;
  int i, r;

  /* Defaults */
  params.p_prefix[0] = 0;
  params.p_uid = 0;
  params.p_gid = 0;
  params.p_file_mask = 0755;
  params.p_dir_mask = 0755;
  params.p_case_insens = FALSE;

  /* If we have been given an options string, parse options from there. */
  for (i = 1; i < env_argc - 1; i++)
	if (!strcmp(env_argv[i], "-o"))
		optset_parse(optset_table, env_argv[++i]);

  /* Initialize the HGFS library. If this fails, exit immediately. */
  if ((r = hgfs_init(&table)) != OK) {
	if (r == EAGAIN)
		printf("HGFS: shared folders are disabled\n");
	else
		printf("HGFS: unable to initialize HGFS library (%d)\n", r);

	return r;
  }

  /* Now initialize the SFFS library. */
  if ((r = sffs_init("HGFS", table, &params)) != OK) {
	hgfs_cleanup();

	return r;
  }

  return OK;
}

/*===========================================================================*
 *				sef_local_startup			     *
 *===========================================================================*/
static void sef_local_startup(void)
{
/* Local SEF initialization.
 */

  /* Register init callback. */
  sef_setcb_init_fresh(sef_cb_init_fresh);

  /* Register signal callback. SFFS handles this. */
  sef_setcb_signal_handler(sffs_signal);

  sef_startup();
}

/*===========================================================================*
 *				main					     *
 *===========================================================================*/
int main(int argc, char **argv)
{
/* The main function of this file server.
 */

  env_setargs(argc, argv);
  sef_local_startup();

  sffs_loop();

  hgfs_cleanup();

  return 0;
}
